backend/api_calls.py

```diff
@@ -877,6 +877,168 @@ def RandomWords(n):
+  if 'groupId' in request:
+    valid_args.append('groupId')
+  elif 'playerId' in request:
+    valid_args.append('playerId')
+  else:
+    raise InvalidInputError('Must include either a playerId or a groupId')
+
+  required_args = list(valid_args)
+  required_args.extend(['message', 'app', 'vibrate', 'sound', 'destination',
+                        'sendTime', 'groupId', 'icon'])
+  ValidateInputs(request, firebase, required_args, valid_args)
+
+  if 'previewMessage' not in request:
+    request['previewMessage'] = textwrap.wrap(request['message'], 100)[0]
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon', 'previewMessage']
+
+  for property in properties:
+    if property in request:
+      put_data[property] = request[property]
+
+  return firebase.put('/notifications',
+                      request['notificationId'], put_data)
+
+
+def UpdateNotification(request, firebase):
+  """Update a queued notification.
+
+  Validation:
+    notificationId exists.
+
+  Args:
+    notificationId: Queued Notification id to create.
+    message: Message to send to client.
+    previewMessage: Short preview of the message conents. Optional.
+    app: Whether to send as a push notification to the iOS/Android/Web clients.
+    vibrate: Whether the notification should vibrate on iOS/Android.
+    sound: Whether the notification should play a sound on iOS/Android.
+    destination: URL that notification should open when clicked.
+    sendTime: When to spawn notifications from this template. UNIX timestamp.
+    playerId: Player id to send message to.
+    groupId: Who to send the notifications to.
+    icon: An icon code to show.
+
+  Firebase entries:
+    /notifications/%(notificationId)
+"""
+  valid_args = ['notificationId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+
+  put_data = {}
+  properties = ['message', 'app', 'vibrate', 'sound', 'destination', 'sendTime',
+                'groupId', 'playerId', 'icon']
+
+  for property in properties:
+    if property in request:
+      put_data[property] = request[property]
+
+  return firebase.patch('/notifications/%s' % request['notificationId'], put_data)
+
+
+def MarkNotificationSeen(request, firebase):
+  """Updates the notification's seenTime.
+
+  Validation:
+    notificationId must exist.
+
+  Args:
+    notificationId: Notification Id to update.
+
+  Firebase entries:
+    /notifications/notifications/%(notificationId)
+"""
+  valid_args = ['notificationId', 'playerId']
+  required_args = list(valid_args)
+  ValidateInputs(request, firebase, required_args, valid_args)
+  put_data = {
+    'seenTime': time.time()
+  }
+  results = []
+  results.append(firebase.patch('/notifications/notifications/%s' % (
+      request['notificationId']), put_data))
+  return results
+
+
+def _handleNotification(self, firebase, notification_id, notification):
```

Comment:
Whoops, copied out of a notifications class I made before I learned about Google App Engine Cron. Cleaned up lingering self reference.